### DISCUSSION TITLE
Discussion: Using Statement And IDisposable

### DISCUSSION THREAD

**Discussion by ajdepersio** [dotnet/csharplang#2948] [General]
[using](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-statement)
[IDisposable](https://docs.microsoft.com/en-us/dotnet/api/system.idisposable?view=netframework-4.8)

The current implementation of the using statement and accompanying documentation can lead to conflicting opinions on what is proper usage and what is a misuse.

At its core, using(IDisposable) is a shorthand of the try { } finally { Dispose(); } pattern that was common prior to its introduction.  The intention of the IDisposable interface is to [release unmanaged resources](https://docs.microsoft.com/en-us/dotnet/api/system.idisposable?view=netframework-4.8#remarks).  In practice, the using statement is often used as a scope-based callback.  As a result, the IDisposable interface is often used in classes which are not dealing with unmanaged resources which can be considered a misuse of the interface.

There's evidence to support that this was intentional back when the using statement was introduced: [(Using for purposes other than disposable objects…)](https://blogs.msdn.microsoft.com/ericgu/2004/03/28/using-for-purposes-other-than-disposable-objects/) and there is even examples of such within Microsoft's own code: [MvcForm.cs](https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.ViewFeatures/src/Rendering/MvcForm.cs).

I'm of the belief that this is not a misuse, but ideally there would be another interface that could be used within a using statement or a mechanism for defining a callback function when the using statement's scope ends.  At minimum, I would like to see clarification on the IDisposable's documentation.

**Comment by Thaina** [comment]
Disagree

**_The primary use_** is not one and only of **_The intention_** of `using` and `IDisposable`

It just being _The primary use_. Secondary is everything else we would wanted to

**Comment by ajdepersio** [comment]
> Disagree
> 
> **_The primary use_** is not one and only of **_The intention_** of `using` and `IDisposable`
> 
> It just being _The primary use_. Secondary is everything else we would wanted to

Right, I agree the wording does indicate there are other valid uses, but If that is the official stance on the interface then the documentation should be updated to make that more clear.

**Comment by franchyd** [comment]
@Thaina I mean this direct quote[ "You should implement IDisposable only if your type uses unmanaged resources directly."](https://docs.microsoft.com/en-us/dotnet/api/system.idisposable?view=netframework-4.8#implementing-idisposable)  Kind-of indicates otherwise.

[Edit] It seems the link brings you to the wrong place on the page sometimes, find the section titled "Implementing IDisposable"

**Comment by 333fred** [comment]
I'm confused what the ask here is. Are you asking for the documentation on IDisposable to be updated, a new interface to be introduced, or something else?

**Comment by franchyd** [comment]
@333fred I interpret it as asking for either documentation update reflecting that IDisposable is not just for unmanaged resources anymore, or a new interface that is compatible with using statements that can be used in place of IDisposable when not using unmanaged resources.

**Comment by ajdepersio** [comment]
@333fred
See @DanFTRX comment. 

A third, though less desirable, option would be that the documentation for IDisposable is updated so that it's explicitly clear that it's only for unmanaged resources and to use the try {}  finally {} pattern for classes working with managed resources that you want to "Dispose" of.

**Comment by canton7** [comment]
People already use IDisposable for purposes other than unmanaged resources. Introducing a new interface here just means that people are going to use both IDisposable and the new interface, creating a mess.

Let's just update the docs to reflect what people actually use it for.

Let's not also forget that the "modern" (post-.NET 2.0) recommended way of handling unmanaged resources is using a `SafeHandle`. The current advice should be *not* to use IDisposable for unmanaged resources. (Excepting performance-critical scenarios and people who know the pitfalls of releasing unmanaged resources from finalizers). 

Consumers of a class - i.e. the code which has to use a `using` statement, or some other hypothetical new construct - doesn't care whether the class is releasing unmanaged resources or doing something else. They shouldn't need to care.

**Comment by dmitriyse** [comment]
I have created proposal related to this discussion. See, #3025
